### ISSUE TITLE
Regression TypeAliasType not resolved since 1.1.370 

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#9026]
**Describe the bug**
I realized that some type statements became incorrect and was able to track it down that it changed in 1.1.370.
EDIT: Crucial is that the `TypeAliasType` definition does not happen on the top-level but in a function.

I would expect for the type-hint in the function to be the same as the one defined globally below as it was prior to 1.1.370.

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.381&pythonVersion=3.11&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAKopJgo0AySAzjDQMrEwBQoSLESo0AfWIAPGMRTtG8zNlz4iCYgEEANknLt1xfvwDEUAArgARtuIRyMJAGMAXMYJQAvIRIUQACgAiAkCASn4AE2JgWGJOcTQ%2BJFkIf3ZibWBQtyhcqDMAWiKiqAA5XHttKD8oYqL%2BPKg2TgB5EF4YD29DHT0DEiDmmDaOkJp6RQBtIcmCAF0ePlm5hZENcQRKcgh2T38CKlDwxvYAVyt2JxAkBFkIryaOYfa%2BAknUGDncswIwMG1HAgXFAAIwAOnBAGYAGwATmBQxGr3eKE%2BAG4oLgoJoUHBMCgoNoHMQQFAAG4khRMdgNPIgYgU8jacTwDRpc6Xa63YgRY55MzgqEAFkhBRh8KggVZxGmTxRny%2BAB81jKOvKVoFad9QRCwZCAOwABmBOLgWvyOqhAA4QSbcWCoP5yrJgewABZgADu7CgAHE5CTnL19PioBA8cBTtoquxtghbOFjGYWgBpfi%2B7RgKxM4M%2B7okXOGIIZrM53T6QLjBhMWWcZaLGDLVbSjZbHZ7A5HfhnC5XG53B4l7PaXPq-j0xnM6Xs3tcu7hMw6bTA6W1xsfFZQZWrtUbub8IA)

```python
from typing import TypeVar, Union, List, Set
from typing_extensions import TypeAliasType

# Problematic:

T = TypeVar("T")
def test_getitem(self):
    # ----- Normal Var -----
    ListOrSetT = TypeAliasType("ListOrSetT", Union[List[T], Set[T]], type_params=(T,))
    subscripted = ListOrSetT[int]  # Tooltip: 1.1.369: ListOrSetT[int]; or Any in later versions
    reveal_type(subscripted)
    # 1.1.343-369: "type[List[int]] | type[Set[int]]"
    # 1.1.370: Any
    # 1.1.381: Any. (Note: shows GenericAlias in my actual sample)

# OK
GlobalAlias = TypeAliasType("GlobalAlias", Union[List[T], Set[T]], type_params=(T,))
subscripted = GlobalAlias[int]
reveal_type(subscripted)
# All: type[List[int]] | type[Set[int]]

```

**VS Code extension or command-line**
1.1.381

**Comment by erictraut** [comment]
In your code sample, `T` isn't defined. I presume that it's a TypeVar?

The `reveal_type` call is intended to reveal the type of a value expression. For example, `reveal_type(1)` reveals the type `Literal[1]`. You're using it here for the expression `ListOrSetT[int]`, which would normally be used as a type expression. This expression can be used as a value expression as well, but its runtime type is not the same as the type it describes when used as a type expression. I suspect that's what's confusing you. Generally, it doesn't make sense to use `reveal_type` on a type alias because it's not that interesting to ask "what is the runtime type of the object that represents this type alias?".

In any case, I don't see any bug here. If I'm missing something, please explain.

If you're interested in understanding more about the difference between type expressions and value expressions, you may find [this section of the Typing Spec](https://typing.readthedocs.io/en/latest/spec/annotations.html) useful. Also, [draft PEP 747](https://discuss.python.org/t/pep-747-typeexpr-type-hint-for-a-type-expression/55984/44) might be of interest to you.

**Comment by Daraan** [comment]
Thank you for the answers and Sorry for the confusion.

I wrote this issue because I noticed that the tooltips by pylance changed compared to a few weeks ago. From what I see do the tooltips always align with the `reveal_type` result by pyright, hence putting `reveal_type` in the code as to more easily check it via CLI (*pyright-python*).

Maybe my wording with "regression" is wrong here? Reverting from *useful type-hint* -> `Any` -> `GenericAlias` felt more like a regression to me.

Alternatively, you could see this issue as an "enhancement request" to bring back the old behavior b) or an alternative type-hint  as the current tooltip/revealed_type "`GenericAlias`" is not really informative.

---

I think it would be more useful if the statically revealed type of
```
reveal_type(ListOrSetT[int])  # Currently: GenericAlias
```
would be either
a) The alias created by `repr`, i.e. alias[+parameters] like `ListOrSetT[int]`
b) The old `TypeAliasType.__value__` behavior `type[List[int]] | type[Set[int]]`

**Comment by Daraan** [comment]
Thank you very much for the answer and explanation.

I am very sorry for not providing you a complete working example. I did not properly mirror the original source code with the error. The crucial difference is that the definition does not happen in the top-level but a function.
I have modified the code above and provided it in the [playground](https://pyright-play.net/?pyrightVersion=1.1.381&pythonVersion=3.11&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAKopJgo0AySAzjDQMrEwBQoSLESo0AfWIAPGMRTtG8zNlz4iCYgEEANknLt1xfvwDEUAArgARtuIRyMJAGMAXMYJQAvIRIUQACgAiAkCASn4AE2JgWGJOcTQ%2BJFkIf3ZibWBQtyhcqDMAWiKiqAA5XHttKD8oYqL%2BPKg2TgB5EF4YD29DHT0DEiDmmDaOkJp6RQBtIcmCAF0ePlm5hZENcQRKcgh2T38CKlDwxvYAVyt2JxAkBFkIryaOYfa%2BAknUGDncswIwMG1HAgXFAAIwAOnBAGYAGwATmBQxGr3eKE%2BAG4oLgoJoUHBMCgoNoHMQQFAAG4khRMdgNPIgYgU8jacTwDRpc6Xa63YgRY55MzgqEAFkhBRh8KggVZxGmTxRny%2BAB81jKOvKVoFad9QRCwZCAOwABmBOLgWvyOqhAA4QSbcWCoP5yrJgewABZgADu7CgAHE5CTnL19PjYG7iFByE4YKcmVB2NsELZwsYzC0ANL8X3aMBWJnBn3dEgFwxBbO5-O6fSBcYMJiyzjLRYwZaraUbLY7PYHI78M4XK43O4Pct57QF9X8emM5nS9kDrl3cJmHTaYHShstj4rKDKjdq7dzfhAA)
